File: openclaw-plugin/src/plugin/command-handler.js
// Shared command handler logic — used by both CJS (index.js) and ESM (entry.mjs) plugin entries
// Extracted to eliminate duplication of formatToolInput, storeSessionRoute, setupProcessForwarding,
// and all 10 command handlers.

const path = require('path');
const { GitSnapshot } = require('../core/git-snapshot');
const { ContextManager } = require('../core/context-manager');

/**
 * Format a tool's raw JSON input into a human-readable preview string.
 * @param {string} toolName - Tool name (Bash, Write, Edit, etc.)
 * @param {string} rawInput - Raw JSON string of the tool input
 * @returns {string} Truncated preview of the tool input
 */
function formatToolInput(toolName, rawInput) {
  try {
    const input = JSON.parse(rawInput);
    if (toolName === 'Bash') return input.command?.slice(0, 200) || '(no command)';
    if (toolName === 'Write' || toolName === 'Edit') return input.file_path || '(no path)';
    return JSON.stringify(input).slice(0, 200);
  } catch {
    return rawInput?.slice(0, 200) || '(no content)';
  }
}

/**
 * Store routing info for a session so output/error messages can be delivered
 * to the correct user, channel, and account.
 * @param {Map} sessionRoutes - Map of sessionId -> { senderId, channelId, accountId }
 * @param {string} sessionId
 * @param {string} senderId
 * @param {string} channelId
 * @param {string} accountId
 */
function storeSessionRoute(sessionRoutes, sessionId, senderId, channelId, accountId) {
  sessionRoutes.set(sessionId, {
    senderId,
    channelId,
    accountId,
  });
}

/**
 * Forward stdout/stderr from a Claude CLI process to the user via messenger,
 * and clean up the session route on process exit.
 * @param {object} messenger - FeishuMessenger instance
 * @param {object} bridge - ClaudeBridge instance
 * @param {Map} sessionRoutes - Map of sessionId -> route info
 * @param {ChildProcess} proc - The spawned Claude CLI process
 * @param {string} sessionId
 */
function setupProcessForwarding(messenger, bridge, sessionRoutes, proc, sessionId) {
  const route = sessionRoutes.get(sessionId);
  const target = route?.senderId;
  let outputBuffer = '';

  proc.stdout.on('data', (chunk) => {
    outputBuffer += chunk.toString();
    bridge.updateActivity(sessionId);

    const lines = outputBuffer.split('\n');
    if (lines.length > 1) {
      outputBuffer = lines.pop();
      const completeLines = lines.join('\n');
      if (completeLines.trim() && target) {
        messenger.sendToUser(target, completeLines, {
          channelId: route?.channelId,
          accountId: route?.accountId
        }).catch(err => {
          console.error('[cc-bridge] Message delivery failed:', err.message);
        });
      }
    }
  });

  proc.stderr.on('data', (chunk) => {
    const text = chunk.toString().trim();
    if (text && target) {
      messenger.sendToUser(target, messenger.formatErrorMessage(text), {
        channelId: route?.channelId,
        accountId: route?.accountId
      }).catch(err => {
        console.error('[cc-bridge] Message delivery failed:', err.message);
      });
    }
  });

  proc.on('exit', (code) => {
    if (outputBuffer.trim() && target) {
      messenger.sendToUser(target, outputBuffer.trim(), {
        channelId: route?.channelId,
        accountId: route?.accountId
      }).catch(err => {
        console.error('[cc-bridge] Message delivery failed:', err.message);
      });
      outputBuffer = '';
    }

    if (code !== 0 && code !== null && target) {
      messenger.sendToUser(target, `会话进程退出 (code: ${code})`, {
        channelId: route?.channelId,
        accountId: route?.accountId
      }).catch(err => {
        console.error('[cc-bridge] Message delivery failed:', err.message);
      });
    }

    sessionRoutes.delete(sessionId);
  });
}

/**
 * Handle a /cc command and return a text reply.
 *
 * @param {object} deps - All runtime dependencies
 * @param {object} deps.bridge          - ClaudeBridge instance
 * @param {object} deps.approvalServer  - ApprovalServer instance (may be null before gateway_start)
 * @param {object} deps.sessionManager  - PersistentSessionManager instance
 * @param {object} deps.messenger       - FeishuMessenger instance
 * @param {object} deps.commandParser   - CommandParser instance
 * @param {Map}    deps.contextManagers - Map<workspace, ContextManager>
 * @param {Map}    deps.sessionRoutes   - Map<sessionId, route>
 * @param {object} deps.hookInbox       - HookInbox instance (may be null before gateway_start)
 * @param {object} deps.pluginConfig    - Plugin config object
 * @param {string} deps.defaultWorkspace - Default workspace path
 * @param {string} deps.currentMode     - 'efficient' or 'strict' (mutable via deps reference)
 * @param {string} commandName - One of: cc, cc_start, cc_stop, cc_status, cc_answer, cc_approve, cc_deny, cc_revert, cc_context, cc_mode
 * @param {string} args       - Command arguments (everything after the command name)
 * @param {string} senderId   - User who sent the command
 * @param {string} channelId  - Channel where the command was sent
 * @param {string} accountId  - Account context
 * @returns {Promise<string>} Reply text
 */
async function handleCommand(deps, commandName, args, senderId, channelId, accountId) {
  const {
    bridge,
    approvalServer,
    sessionManager,
    messenger,
    commandParser,
    contextManagers,
    sessionRoutes,
    hookInbox,
    pluginConfig,
    defaultWorkspace,
    currentMode,
  } = deps;

  // /cc <prompt> — one-shot task
  if (commandName === 'cc') {
    const prompt = args.trim();
    if (!prompt) return '用法: /cc <任务描述>';

    const workspace = defaultWorkspace;
    const existingSessionId = bridge.findActiveSession(senderId);
    if (existingSessionId) {
      const proc = bridge.processMap.get(existingSessionId);
      if (proc && proc.exitCode === null) {
        proc.stdin.write(prompt + '\n');
        bridge.updateActivity(existingSessionId);
        return '已发送到持久会话。';
      }
    }

    const { sessionId } = await bridge.spawnSession(senderId, workspace, prompt);
    storeSessionRoute(sessionRoutes, sessionId, senderId, channelId, accountId);
    const proc = bridge.processMap.get(sessionId);
    if (proc) setupProcessForwarding(messenger, bridge, sessionRoutes, proc, sessionId);

    return `任务已提交 (ID: ${sessionId.slice(0, 8)})`;
  }

  // /cc_start — start persistent session
  if (commandName === 'cc_start') {
    const workspace = defaultWorkspace;
    const existing = bridge.findActiveSession(senderId);
    if (existing) {
      const meta = bridge.sessionMeta.get(existing);
      const proc = bridge.processMap.get(existing);
      return messenger.formatSessionStatus(meta, proc, existing);
    }

    const { sessionId } = await bridge.spawnSession(senderId, workspace, '');
    storeSessionRoute(sessionRoutes, sessionId, senderId, channelId, accountId);
    sessionManager.activate(senderId, workspace, sessionId);

    const proc = bridge.processMap.get(sessionId);
    if (proc) setupProcessForwarding(messenger, bridge, sessionRoutes, proc, sessionId);

    let cm = contextManagers.get(workspace);
    if (!cm) {
      cm = new ContextManager(workspace);
      contextManagers.set(workspace, cm);
    }
    cm.cleanOrphanedRules(bridge);
    cm.injectRules();

    const gitSnapshotEnabled = pluginConfig.gitSnapshotEnabled !== false;
    const snapshot = new GitSnapshot(bridge, sessionId);
    const snapshotCreated = gitSnapshotEnabled ? snapshot.create() : false;

    const port = approvalServer?.getPort();
    if (port && hookInbox) {
      hookInbox.writeHookConfig(
        path.join(workspace, '.claude', 'settings.local.json'),
        port,
        currentMode === 'efficient' ? 'Bash' : 'Bash|Write|Edit'
      );
    }

    return `持久会话已启动 (ID: ${sessionId.slice(0, 8)})\n模式: ${currentMode}\n审批服务端口: ${port}\n快照: ${snapshotCreated ? '已创建' : '未创建（非Git目录）'}`;
  }

  // /cc_stop — stop persistent session
  if (commandName === 'cc_stop') {
    const sessionId = bridge.findActiveSession(senderId);
    if (!sessionId || !bridge.sessionMeta.get(sessionId)?.active) {
      return '当前没有持久会话。';
    }

    const meta = bridge.sessionMeta.get(sessionId);
    if (meta.senderId !== senderId) {
      return '无权停止该会话，只有会话创建者可以停止。';
    }

    bridge.terminateSession(sessionId);
    sessionManager.deactivate(senderId);
    sessionRoutes.delete(sessionId);

    const snapshot = new GitSnapshot(bridge, sessionId);
    if (meta.stashRef) snapshot.dropStash();

    const cm = contextManagers.get(meta.cwd);
    if (cm) {
      cm.cleanup();
      contextManagers.delete(meta.cwd);
    }

    GitSnapshot.cleanupOldStashes(bridge, meta.cwd);

    const runtime = Math.round((Date.now() - new Date(meta.startedAt).getTime()) / 60000);
    return `会话已停止 (ID: ${sessionId.slice(0, 8)})\n运行时长: ${runtime} 分钟\n消息数: ${meta.messageCount}`;
  }

  // /cc_status — check session status
  if (commandName === 'cc_status') {
    const sessionId = bridge.findActiveSession(senderId);
    if (!sessionId) return '当前没有活跃的持久会话。';
    const meta = bridge.sessionMeta.get(sessionId);
    const proc = bridge.processMap.get(sessionId);
    return messenger.formatSessionStatus(meta, proc, sessionId);
  }

  // /cc_answer — reply to Claude's question
  if (commandName === 'cc_answer') {
    const answer = args.trim();
    if (!answer) return '用法: /cc_answer <回答内容>';

    const sessionId = bridge.findActiveSession(senderId);
    if (!sessionId) return '没有活跃会话。';

    const proc = bridge.processMap.get(sessionId);
    if (proc && proc.exitCode === null) {
      proc.stdin.write(answer + '\n');
      bridge.updateActivity(sessionId);
      return '已发送回答。';
    }
    return '会话进程已退出，无法回答。';
  }

  // /cc_approve — approve a pending request
  if (commandName === 'cc_approve') {
    const shortId = args.trim();
    if (!shortId) return '用法: /cc_approve <审批ID>';

    const item = approvalServer.store.findByShortId(shortId);
    if (!item) return '未找到该审批请求。';
    if (item.ambiguous) return `ID "${shortId}" 匹配到多个请求，请使用更长的ID。\n匹配: ${item.matches.map(m => m.id.slice(0, 12)).join(', ')}`;
    if (item.status !== 'PENDING') return `该请求已处理: ${item.status}`;

    const meta = bridge.sessionMeta.get(item.sessionId);
    if (meta && meta.senderId !== senderId) {
      return '无权审批该请求，只有会话创建者可以操作。';
    }

    approvalServer.store.resolve(item.id, 'APPROVED');
    return `已批准 #${shortId}: ${item.toolName}`;
  }

  // /cc_deny — deny a pending request
  if (commandName === 'cc_deny') {
    const shortId = args.trim();
    if (!shortId) return '用法: /cc_deny <审批ID>';

    const item = approvalServer.store.findByShortId(shortId);
    if (!item) return '未找到该审批请求。';
    if (item.ambiguous) return `ID "${shortId}" 匹配到多个请求，请使用更长的ID。\n匹配: ${item.matches.map(m => m.id.slice(0, 12)).join(', ')}`;
    if (item.status !== 'PENDING') return `该请求已处理: ${item.status}`;

    const meta = bridge.sessionMeta.get(item.sessionId);
    if (meta && meta.senderId !== senderId) {
      return '无权审批该请求，只有会话创建者可以操作。';
    }

    approvalServer.store.resolve(item.id, 'DENIED');
    return `已拒绝 #${shortId}: ${item.toolName}`;
  }

  // /cc_revert — rollback code changes
  if (commandName === 'cc_revert') {
    const revertArgs = args.trim();

    if (revertArgs === '--confirm') {
      const sessionId = bridge.findActiveSession(senderId);
      if (!sessionId) return '没有活跃的会话。';

      const meta = bridge.sessionMeta.get(sessionId);
      if (!meta?.stashRef) return '无可用的快照，无法回滚。（可能已回滚过或启动时未创建快照）';

      const snapshot = new GitSnapshot(bridge, sessionId);
      const result = snapshot.revert();
      return result.success ? '已回滚到任务前状态。' : `回滚失败: ${result.message}`;
    }

    if (revertArgs === '--cancel') return '已取消回滚。';

    const sessionId = bridge.findActiveSession(senderId);
    if (!sessionId) return '没有活跃的会话。';

    const meta = bridge.sessionMeta.get(sessionId);
    if (!meta?.stashRef) return '无可用的快照。（可能已回滚过或启动时未创建快照）';

    return `确认回滚？\n将恢复到上次 CC 任务前的状态。\n\n确认: /cc_revert --confirm\n取消: /cc_revert --cancel`;
  }

  // /cc_context — view project context
  if (commandName === 'cc_context') {
    const sessionId = bridge.findActiveSession(senderId);
    if (!sessionId) return '没有活跃会话。';

    const meta = bridge.sessionMeta.get(sessionId);
    const cm = contextManagers.get(meta.cwd);
    const contextInfo = cm?.buildContextPrompt(meta.cwd) || '(无上下文)';
    return `当前项目上下文:\n\n${contextInfo}`;
  }

  // /cc_mode — switch approval mode
  if (commandName === 'cc_mode') {
    const mode = args.trim();
    if (!mode || (mode !== 'efficient' && mode !== 'strict')) {
      return `当前模式: ${currentMode}\nefficient — Edit/Write 免审批，仅 Bash 需审批\nstrict — 全部操作需审批\n\n切换: /cc_mode efficient 或 /cc_mode strict`;
    }

    const matcher = mode === 'efficient' ? 'Bash' : 'Bash|Write|Edit';
    if (hookInbox) hookInbox.updateMatcher(matcher);
    if (approvalServer) approvalServer.setMode(mode);
    // Mutate via deps reference so caller sees the update
    deps.currentMode = mode;
    return `已切换到 ${mode} 模式。`;
  }

  return `未知命令: ${commandName}\n\n${commandParser.getHelpText()}`;
}

module.exports = { formatToolInput, storeSessionRoute, setupProcessForwarding, handleCommand };

File: hermes-plugin/core/command_handler.py
"""CommandHandler — handles all 10 CC Bridge commands.

Port of the handleCommand logic from openclaw-plugin/src/plugin/index.js.
All 10 command handlers produce the same response text as the Node.js version.

See: cc-bridge-v3-final-plan.md Sections 7, 8
"""

import json
import logging
import os
import time
from datetime import datetime, timezone
from typing import Any, Optional

from .approval_server import ApprovalServer
from .claude_bridge import ClaudeBridge
from .context_manager import ContextManager
from .git_snapshot import GitSnapshot
from .messenger import DefaultMessenger
from .session_manager import PersistentSessionManager

logger = logging.getLogger(__name__)


def formatToolInput(tool_name: str, raw_input: str) -> str:
    """Format tool input for display in approval notifications."""
    try:
        input_data = json.loads(raw_input)
        if tool_name == "Bash":
            return (input_data.get("command") or "(no command)")[:200]
        if tool_name in ("Write", "Edit"):
            return input_data.get("file_path") or "(no path)"
        return json.dumps(input_data)[:200]
    except (json.JSONDecodeError, TypeError):
        return (raw_input or "(no content)")[:200]


class CommandHandler:
    """Handle all CC Bridge slash commands."""

    def __init__(
        self,
        bridge: ClaudeBridge,
        approval_server: Optional[ApprovalServer],
        session_manager: PersistentSessionManager,
        messenger: DefaultMessenger,
        context_managers: dict[str, ContextManager],
        config: dict[str, Any],
    ) -> None:
        self.bridge = bridge
        self.approval_server = approval_server
        self.session_manager = session_manager
        self.messenger = messenger
        self.context_managers = context_managers
        self.config = config
        self.current_mode: str = config.get("defaultMode", "efficient")
        self.session_routes: dict[str, dict[str, str]] = {}

    def storeSessionRoute(
        self,
        session_id: str,
        sender_id: str,
        channel_id: str,
        account_id: str,
    ) -> None:
        """Store the routing info (sender, channel, account) for a session."""
        self.session_routes[session_id] = {
            "senderId": sender_id,
            "channelId": channel_id,
            "accountId": account_id,
        }

    def setupProcessForwarding(
        self, proc: Any, session_id: str
    ) -> None:
        """Set up stdout/stderr forwarding from the Claude process to the user.

        Reads from the process output stream and forwards lines
        via the messenger to the original sender.
        """
        route = self.session_routes.get(session_id)
        target = route.get("senderId") if route else None

        async def _forward_stdout() -> None:
            output_buffer = ""
            try:
                while True:
                    chunk = await proc.stdout.read(4096)
                    if not chunk:
                        break
                    output_buffer += chunk.decode("utf-8", errors="replace")
                    self.bridge.updateActivity(session_id)

                    lines = output_buffer.split("\n")
                    if len(lines) > 1:
                        output_buffer = lines.pop()
                        complete_lines = "\n".join(lines)
                        if complete_lines.strip() and target:
                            await self.messenger.send_to_user(
                                target, complete_lines,
                                {
                                    "channelId": route.get("channelId") if route else None,
                                    "accountId": route.get("accountId") if route else None,
                                },
                            )
            except Exception:
                logger.warning("stdout forwarding failed", exc_info=True)

            # Flush remaining buffer on exit
            if output_buffer.strip() and target:
                await self.messenger.send_to_user(
                    target, output_buffer.strip(),
                    {
                        "channelId": route.get("channelId") if route else None,
                        "accountId": route.get("accountId") if route else None,
                    },
                )

        async def _forward_stderr() -> None:
            try:
                while True:
                    chunk = await proc.stderr.read(4096)
                    if not chunk:
                        break
                    text = chunk.decode("utf-8", errors="replace").strip()
                    if text and target:
                        await self.messenger.send_to_user(
                            target,
                            self.messenger.formatErrorMessage(text),
                            {
                                "channelId": route.get("channelId") if route else None,
                                "accountId": route.get("accountId") if route else None,
                            },
                        )
            except Exception:
                logger.warning("stderr forwarding failed", exc_info=True)

        async def _on_exit() -> None:
            await proc.wait()
            exit_code = proc.returncode
            if exit_code not in (0, None) and target:
                await self.messenger.send_to_user(
                    target,
                    f"会话进程退出 (code: {exit_code})",
                    {
                        "channelId": route.get("channelId") if route else None,
                        "accountId": route.get("accountId") if route else None,
                    },
                )
            self.session_routes.pop(session_id, None)

        # Schedule forwarding tasks
        import asyncio
        asyncio.create_task(_forward_stdout())
        asyncio.create_task(_forward_stderr())
        asyncio.create_task(_on_exit())

    async def handleCommand(
        self,
        command_name: str,
        args: str,
        sender_id: str,
        channel_id: str,
        account_id: str,
    ) -> str:
        """Dispatch and handle a CC Bridge command. Returns the response text."""

        # /cc <prompt> — one-shot task
        if command_name == "cc":
            return await self._handleCc(args, sender_id, channel_id, account_id)

        # /cc_start — start persistent session
        if command_name == "cc_start":
            return await self._handleCcStart(sender_id, channel_id, account_id)

        # /cc_stop — stop persistent session
        if command_name == "cc_stop":
            return await self._handleCcStop(sender_id)

        # /cc_status — check session status
        if command_name == "cc_status":
            return self._handleCcStatus(sender_id)

        # /cc_answer — reply to Claude's question
        if command_name == "cc_answer":
            return await self._handleCcAnswer(args, sender_id)

        # /cc_approve — approve a pending request
        if command_name == "cc_approve":
            return self._handleCcApprove(args, sender_id)

        # /cc_deny — deny a pending request
        if command_name == "cc_deny":
            return self._handleCcDeny(args, sender_id)

        # /cc_revert — rollback code changes
        if command_name == "cc_revert":
            return await self._handleCcRevert(args, sender_id)

        # /cc_context — view project context
        if command_name == "cc_context":
            return await self._handleCcContext(sender_id)

        # /cc_mode — switch approval mode
        if command_name == "cc_mode":
            return self._handleCcMode(args)

        return f"未知命令: {command_name}"

    # ---- Individual command handlers ----

    async def _handleCc(
        self, args: str, sender_id: str, channel_id: str, account_id: str
    ) -> str:
        """Handle /cc <prompt> — one-shot task."""
        prompt = args.strip()
        if not prompt:
            return "用法: /cc <任务描述>"

        workspace = self.config.get("workspace", os.getcwd())
        existing_session_id = self.bridge.findActiveSession(sender_id)
        if existing_session_id:
            proc = self.bridge.process_map.get(existing_session_id)
            if proc and proc.returncode is None:
                proc.stdin.write(prompt.encode("utf-8") + b"\n")
                await proc.stdin.drain()
                self.bridge.updateActivity(existing_session_id)
                return "已发送到持久会话。"

        result = await self.bridge.spawnSession(sender_id, workspace, prompt)
        session_id = result["sessionId"]
        self.storeSessionRoute(session_id, sender_id, channel_id, account_id)
        proc = self.bridge.process_map.get(session_id)
        if proc:
            self.setupProcessForwarding(proc, session_id)

        return f"任务已提交 (ID: {session_id[:8]})"

    async def _handleCcStart(
        self, sender_id: str, channel_id: str, account_id: str
    ) -> str:
        """Handle /cc_start — start persistent session."""
        workspace = self.config.get("workspace", os.getcwd())
        existing = self.bridge.findActiveSession(sender_id)
        if existing:
            meta = self.bridge.session_meta.get(existing)
            proc = self.bridge.process_map.get(existing)
            return self.messenger.formatSessionStatus(meta, proc, existing)

        result = await self.bridge.spawnSession(sender_id, workspace, "")
        session_id = result["sessionId"]
        self.storeSessionRoute(session_id, sender_id, channel_id, account_id)
        self.session_manager.activate(sender_id, workspace, session_id)

        proc = self.bridge.process_map.get(session_id)
        if proc:
            self.setupProcessForwarding(proc, session_id)

        context_manager = self.context_managers.get(workspace)
        if not context_manager:
            context_manager = ContextManager(workspace)
            self.context_managers[workspace] = context_manager
        context_manager.cleanOrphanedRules(self.bridge)
        context_manager.injectRules()

        git_snapshot_enabled = self.config.get("gitSnapshotEnabled", True)
        snapshot = GitSnapshot(self.bridge, session_id)
        snapshot_created = (
            await snapshot.create() if git_snapshot_enabled else False
        )

        port = self.approval_server.getPort() if self.approval_server else None

        return (
            f"持久会话已启动 (ID: {session_id[:8]})\n"
            f"模式: {self.current_mode}\n"
            f"审批服务端口: {port}\n"
            f"快照: {'已创建' if snapshot_created else '未创建（非Git目录）'}"
        )

    async def _handleCcStop(self, sender_id: str) -> str:
        """Handle /cc_stop — stop persistent session."""
        session_id = self.bridge.findActiveSession(sender_id)
        if not session_id or not self.bridge.session_meta.get(session_id, {}).get("active"):
            return "当前没有持久会话。"

        meta = self.bridge.session_meta.get(session_id)
        if meta and meta.get("senderId") != sender_id:
            return "无权停止该会话，只有会话创建者可以停止。"

        self.bridge.terminateSession(session_id)
        self.session_manager.deactivate(sender_id)
        self.session_routes.pop(session_id, None)

        snapshot = GitSnapshot(self.bridge, session_id)
        if meta and meta.get("stashRef"):
            await snapshot.dropStash()

        cwd = meta.get("cwd") if meta else None
        cm = self.context_managers.get(cwd) if cwd else None
        if cm:
            cm.cleanup()
            self.context_managers.pop(cwd, None)

        if cwd:
            await GitSnapshot.cleanupOldStashes(self.bridge, cwd)

        runtime = 0
        if meta:
            started_ms = datetime.fromisoformat(meta["startedAt"]).timestamp() * 1000
            runtime = round((time.time() * 1000 - started_ms) / 60000)

        message_count = meta.get("messageCount", 0) if meta else 0
        return (
            f"会话已停止 (ID: {session_id[:8]})\n"
            f"运行时长: {runtime} 分钟\n"
            f"消息数: {message_count}"
        )

    def _handleCcStatus(self, sender_id: str) -> str:
        """Handle /cc_status — check session status."""
        session_id = self.bridge.findActiveSession(sender_id)
        if not session_id:
            return "当前没有活跃的持久会话。"
        meta = self.bridge.session_meta.get(session_id)
        proc = self.bridge.process_map.get(session_id)
        return self.messenger.formatSessionStatus(meta, proc, session_id)

    async def _handleCcAnswer(self, args: str, sender_id: str) -> str:
        """Handle /cc_answer <text> — reply to Claude's question."""
        answer = args.strip()
        if not answer:
            return "用法: /cc_answer <回答内容>"

        session_id = self.bridge.findActiveSession(sender_id)
        if not session_id:
            return "没有活跃会话。"

        proc = self.bridge.process_map.get(session_id)
        if proc and proc.returncode is None:
            proc.stdin.write(answer.encode("utf-8") + b"\n")
            await proc.stdin.drain()
            self.bridge.updateActivity(session_id)
            return "已发送回答。"
        return "会话进程已退出，无法回答。"

    def _handleCcApprove(self, args: str, sender_id: str) -> str:
        """Handle /cc_approve <id> — approve a pending request."""
        short_id = args.strip()
        if not short_id:
            return "用法: /cc_approve <审批ID>"

        if not self.approval_server:
            return "审批服务未启动。"

        item = self.approval_server.store.findByShortId(short_id)
        if not item:
            return "未找到该审批请求。"
        if isinstance(item, dict) and item.get("ambiguous"):
            matches = item.get("matches", [])
            ids = ", ".join(m.get("id", "")[:12] for m in matches)
            return f'ID "{short_id}" 匹配到多个请求，请使用更长的ID。\n匹配: {ids}'
        if item.get("status") != "PENDING":
            return f"该请求已处理: {item.get('status')}"

        meta = self.bridge.session_meta.get(item.get("sessionId"))
        if meta and meta.get("senderId") != sender_id:
            return "无权审批该请求，只有会话创建者可以操作。"

        self.approval_server.store.resolve(item["id"], "APPROVED")
        return f"已批准 #{short_id}: {item.get('toolName')}"

    def _handleCcDeny(self, args: str, sender_id: str) -> str:
        """Handle /cc_deny <id> — deny a pending request."""
        short_id = args.strip()
        if not short_id:
            return "用法: /cc_deny <审批ID>"

        if not self.approval_server:
            return "审批服务未启动。"

        item = self.approval_server.store.findByShortId(short_id)
        if not item:
            return "未找到该审批请求。"
        if isinstance(item, dict) and item.get("ambiguous"):
            matches = item.get("matches", [])
            ids = ", ".join(m.get("id", "")[:12] for m in matches)
            return f'ID "{short_id}" 匹配到多个请求，请使用更长的ID。\n匹配: {ids}'
        if item.get("status") != "PENDING":
            return f"该请求已处理: {item.get('status')}"

        meta = self.bridge.session_meta.get(item.get("sessionId"))
        if meta and meta.get("senderId") != sender_id:
            return "无权审批该请求，只有会话创建者可以操作。"

        self.approval_server.store.resolve(item["id"], "DENIED")
        return f"已拒绝 #{short_id}: {item.get('toolName')}"

    async def _handleCcRevert(self, args: str, sender_id: str) -> str:
        """Handle /cc_revert — rollback code changes."""
        revert_args = args.strip()

        if revert_args == "--confirm":
            session_id = self.bridge.findActiveSession(sender_id)
            if not session_id:
                return "没有活跃的会话。"

            meta = self.bridge.session_meta.get(session_id)
            if not meta or not meta.get("stashRef"):
                return "无可用的快照，无法回滚。（可能已回滚过或启动时未创建快照）"

            snapshot = GitSnapshot(self.bridge, session_id)
            result = await snapshot.revert()
            if result.get("success"):
                return "已回滚到任务前状态。"
            return f"回滚失败: {result.get('message')}"

        if revert_args == "--cancel":
            return "已取消回滚。"

        session_id = self.bridge.findActiveSession(sender_id)
        if not session_id:
            return "没有活跃的会话。"

        meta = self.bridge.session_meta.get(session_id)
        if not meta or not meta.get("stashRef"):
            return "无可用的快照。（可能已回滚过或启动时未创建快照）"

        return (
            "确认回滚？\n"
            "将恢复到上次 CC 任务前的状态。\n\n"
            "确认: /cc_revert --confirm\n"
            "取消: /cc_revert --cancel"
        )

    async def _handleCcContext(self, sender_id: str) -> str:
        """Handle /cc_context — view project context."""
        session_id = self.bridge.findActiveSession(sender_id)
        if not session_id:
            return "没有活跃会话。"

        meta = self.bridge.session_meta.get(session_id)
        if not meta:
            return "没有活跃会话。"

        cm = self.context_managers.get(meta.get("cwd"))
        context_info = (
            await cm.buildContextPrompt(meta["cwd"]) if cm else "(无上下文)"
        )
        return f"当前项目上下文:\n\n{context_info}"

    def _handleCcMode(self, args: str) -> str:
        """Handle /cc_mode [efficient|strict] — switch approval mode."""
        mode = args.strip()
        if not mode or mode not in ("efficient", "strict"):
            return (
                f"当前模式: {self.current_mode}\n"
                "efficient — Edit/Write 免审批，仅 Bash 需审批\n"
                "strict — 全部操作需审批\n\n"
                "切换: /cc_mode efficient 或 /cc_mode strict"
            )

        if self.approval_server:
            self.approval_server.setMode(mode)
        self.current_mode = mode
        return f"已切换到 {mode} 模式。"